### DISCUSSION TITLE
Can SignalR be used to create a server-side browserless websockets client?

### DISCUSSION THREAD

**Discussion by stemixreal** [dotnet/aspnetcore#28531] [Q&A]
Is it possible, using SignalR, to build a server-side wss websockets client that can request a connection then listen for messages? 

Almost all the examples I can find, on various Azure services - application gateway, app service, and others, are for communication between your own webserver and clients in browsers. 

I have no development experience with SignalR, more than happy to learn I'm just trying to save time by asking if it's possible before I dig further. The only example I've been able to find that creates a server-side client is one that runs in a webjob using dotnet and WebSocketSharp (https://github.com/sta/websocket-sharp). It's from 5 years ago though.

Thanks

**Comment by BrennanConroy** [comment]
Take a look at the docs, they might help you https://docs.microsoft.com/aspnet/core/signalr/background-services?view=aspnetcore-5.0

**Reply by stemixreal** [reply]
Hi 

Thanks. Unfortunately not really helpful. Again all examples are focused on hosting the server yourself and/or connecting to SignalR Hubs. That article covers

    Hosting SignalR Hubs using a background worker process hosted with ASP.NET Core.
    Sending messages to connected clients from within a .NET Core BackgroundService.

And contains similar information to other articles in that doc set. Maybe I can do the work in a hub but the examples and documentation don't illustrate that scenario which makes me think it isn't supported.

I need to open secure websocket connections to external services from within azure without any browser involvement. The technology used doesn't matter. I'm just looking for a starting point so I can get on with it. I have no choice but to do it, by the end of the month :) I've read loads of articles but all end up in a dead-end for me. Could be I'm asking the wrong questions. So if you have any pointers to creating server-side websockets clients in azure I'd really appreciate it.

Thanks

**Reply by BrennanConroy** [reply]
> I need to open secure websocket connections to external services

Do these external services implement the SignalR protocol or are they just regular websocket endpoints?

SignalR is a protocol and generally requires a server that hosts the endpoint that your clients connect to and both sides speak the protocol. This means you can't use a SignalR client to connect to a random websocket endpoint unless it speaks that protocol.

If you're looking for a websocket client that you can connect to an endpoint from your server then you should look at [ClientWebSocket](https://docs.microsoft.com/dotnet/api/system.net.websockets.clientwebsocket?view=net-5.0)

**Reply by stemixreal** [reply]
Thanks Brennan, they're regular endpoints. I'll look into ClientWebSocket. I'd also had a look at https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets?view=aspnetcore-5.0. I don't know enough about this yet to know if I can utilize that functionality in a client. I'll get there :) Thanks for your help Brennan
